pkg/noun/jets/c/mas.c

```diff
@@ -1,45 +1,49 @@
-/// @file
-
 #include "jets/q.h"
 #include "jets/w.h"
 
 #include "noun.h"
 
+u3_noun
+u3qc_mas(u3_atom a)
+{
+  c3_w b_w;
 
-  u3_noun
-  u3qc_mas(u3_atom a)
-  {
-    c3_w b_w;
-    u3_atom c, d, e, f;
+  if ( c3y == u3a_is_cat(a) ) {
+    b_w = c3_bits_word(a);
 
-    b_w = u3r_met(0, a);
-    if ( b_w < 2 ) {
+    if ( 2 > b_w ) {
       return u3m_bail(c3__exit);
     }
     else {
-      c = u3qc_bex((b_w - 1));
-      d = u3qc_bex((b_w - 2));
-      e = u3qa_sub(a, c);
-      f = u3qc_con(e, d);
-
-      u3z(c);
-      u3z(d);
-      u3z(e);
-
-      return f;
+      a  &= ~((c3_w)1 << (b_w - 1));
+      a  |=  ((c3_w)1 << (b_w - 2));
+      return a;
     }
   }
-  u3_noun
-  u3wc_mas(u3_noun cor)
-  {
-    u3_noun a;
-
-    if ( (u3_none == (a = u3r_at(u3x_sam, cor))) ||
-         (c3n == u3ud(a)) )
-    {
-      return u3m_bail(c3__exit);
-    } else {
-      return u3qc_mas(a);
+  else {
+    b_w = u3r_met(0, a);
+
+    if ( 64 > b_w ) {
+      c3_d a_d = u3r_chub(0, a);
+      a_d  &= ~((c3_d)1 << (b_w - 1));
+      a_d  |=  ((c3_d)1 << (b_w - 2));
+      return u3i_chub(a_d);
+    }
+    else {
+      u3i_slab sab_u;
+      u3i_slab_from(&sab_u, a, 0, --b_w);
+
+      sab_u.buf_w[(b_w >> 5)] &= ((c3_w)1 << (b_w & 31)) - 1;
```

Comment:
NB: slab_from rounds *up* to word boundaries in the slice specified, so we need to mask off any more significant bits than we want.